### PR TITLE
RecoveryPlugin

### LINKED ISSUE #6: Add simple recovery module
- Allows replacing a specified owner by a non-owner account
- requires root access
- Implement timedelay
  - Reference for timedelay: https://github.com/rmeissner/delayed-tx-module/blob/main/contracts/DelayedTxModule.sol

### PR DESCRIPTION
**Pull request by akshay-ap**
Fixes #6 

Changes in PR:
- [x] Create RecoveryPlugin Contract
- [x] Update deployment script
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: contracts/contracts/RecoveryWithDelayPlugin.sol [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+     * @dev Returns the chain id used by this contract.
+     */
+    function getChainId() public view returns (uint256) {
+        uint256 id;
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            id := chainid()
+        }
+        return id;
+    }
```
Comment (mmv08) [inline contracts/contracts/RecoveryWithDelayPlugin.sol:131]: you can use block.chainid. It should be supported in solidity 0.8.x

File: contracts/contracts/RecoveryWithDelayPlugin.sol [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    struct Announcement {
+        uint64 executionTime; // Block time in seconds when the announced transaction can be executed
+        uint64 validityDuration; // Duration in seconds the announcement is valid after delay is over (0 is valid forever)
+        bool executed; // Flag if the announced transaction was executed
+    }
+
+    // Contract storage
+
+    // Only recoverer can initiate recovery process
+    address public immutable recoverer;
```
Comment (mmv08) [inline contracts/contracts/RecoveryWithDelayPlugin.sol:35]: nitpick: Immutables aren't technically storage, are they?

File: contracts/contracts/RecoveryWithDelayPlugin.sol [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    function executeFromPlugin(
+        ISafeProtocolManager manager,
+        ISafe safe,
+        address prevOwner,
+        address oldOwner,
+        address newOwner,
+        uint256 nonce
+    ) external returns (bytes memory data) {
+        bytes32 txHash = getTransactionHash(address(manager), address(safe), prevOwner, oldOwner, newOwner, nonce);
+        Announcement memory announcement = announcements[txHash];
```
Comment (mmv08) [inline contracts/contracts/RecoveryWithDelayPlugin.sol:93]: Does this actually change the storage? on line `110` you set `executed` to true

File: contracts/contracts/RecoveryWithDelayPlugin.sol [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        address account,
+        address prevOwner,
+        address oldOwner,
+        address newOwner,
+        uint256 nonce,
+        uint64 executionTime,
+        uint64 validityDuration
+    ) external onlyRecoverer {
+        bytes32 txHash = getTransactionHash(manager, account, prevOwner, oldOwner, newOwner, nonce);
+        Announcement memory announcement = announcements[txHash];
```
Comment (mmv08) [inline contracts/contracts/RecoveryWithDelayPlugin.sol:182]: does this write to storagE?

File: contracts/contracts/RecoveryWithDelayPlugin.sol (hunk 2 above)
Comment (akshay-ap) [reply contracts/contracts/RecoveryWithDelayPlugin.sol:35]: yes. should move the comment below.
Initially had thought of keeping `recoverer` updatable.

File: contracts/contracts/RecoveryWithDelayPlugin.sol (hunk 4 above)
Comment (akshay-ap) [reply contracts/contracts/RecoveryWithDelayPlugin.sol:182]: line `182` is intended to only read from storage.

File: contracts/contracts/RecoveryWithDelayPlugin.sol (hunk 3 above)
Comment (akshay-ap) [reply contracts/contracts/RecoveryWithDelayPlugin.sol:92]: The code sets the `storage` at line `110` (now `109`)

File: contracts/contracts/RecoveryWithDelayPlugin.sol [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    /**
+     * @notice Returns the transaction hash for a recovery transaction.
+     * @param manager Address of the manager contract.
+     * @param account Address of the safe account.
+     * @param prevOwner Address of the owner previous to the owner to be replaced in the linked list
+     * @param oldOwner Address of the owner to be replaced.
+     * @param newOwner Address of the new owner.
+     * @param nonce A uint256 used to uniquely identify a recovery transaction.
+     */
+    function getTransactionHashData(
```
Comment (mmv08) [inline contracts/contracts/RecoveryWithDelayPlugin.sol:129]: Nitpick: I'd move this function lower close to getTransactionHash

**Pull request merged**